keps/sig-storage/20200609-csi-service-account-token.md

```diff
@@ -0,0 +1,129 @@
+traits:
+
+1.  It will not work for csi drivers which run as a different non-root user than
+    the pods. See
+    [file permission section for service account token](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/20180515-svcacct-token-volumes.md#file-permission).
+2.  CSI driver will have access to the secrets of pods that do not use it
+    because the CSI driver should have a `hostPath` volume for the `pods`
+    subdirectory to read the token.
+3.  The audience of the token is defaulted to kube apiserver.
+4.  The token is not guaranteed to be available (e.g.
+    `automountServiceAccountToken=false`).
+
+### User stories
+
+- HashiCorp Vault provider for secret store CSI driver requires service
+  account token of the pods they are mounting secrets at to authenticate to
+  Vaults. The provisioned secrets also have given TTL in Vault, so it is
+  neccessary get tokens after the inital mount.
+- Cert manager CSI dirver will create CertificateRequests on behalf of the
+  pods.
+- Amazon EFS CSI driver wants the service account tokens of pods to exchange for
+  AWS credentials.
+
+## Proposal
+
+Kubelet should inject audience-bounded service account tokens requested by the
+CSI driver from the `TokenRequest` API in `VolumeContext` to
+`NodePublishVolume`.
+
+### CSIDriver Object
+
+```go
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+    ... // existing fields
+
+    RequiresRemount *bool
+    ServiceAccountTokenAudience *string
+}
+```
+
+`ServiceAccountTokenAudience` will be used to set on the created token, which
+will be validated to not be empty string (i.e. kube apiserver).
+`RequiresRemount` is optional, which should be only set when the mounted volumes
+by the CRI driver have TTL and require re-validation on the token.
+
+The token will be set in `VolumeContext`:
+
+```go
+"csi.storage.k8s.io/serviceAccount.token": token
+```
+
+With the new fields:
+
+1.  The issued token for CSI driver cannot authenticated to kube apiserver.
+    Storage provider is expected to call TokenReview api where audience is
+    specified to validate the token.
+2.  If storage provider requires periodic authentication, `RequiresRemount` can
+    be used to trigger a sequence of `NodePublishVolume` calls. Each call will
+    have valid token in the `VolumeContext`.
+
+### Example Workflow
+
+Take the Vault provider for secret store CSI driver as an example:
+
+1.  Create `CSIDriver` object with `ServiceAccountTokenAudience=vault` and
+    `RequiresRemount=true`
+2.  When the volume manager of kubelet sees a new volume, the pod object in
+    `mountedPods` will have `requiresRemound=true` after `MarkRemountRequired`
+    is called. `MarkRemountRequired` will call into `RequiresRemount` of the
+    in-tree csi plugin to fetch the `CSIDriver` object.
+3.  Before `NodePublishVolume` call, kubelet will request token from
```

Comment:
Some CSI drivers need ControllerPublish and NodeStage calls. Do you plan to add the token to these too? If not, 
can we safely assume that these calls won't ever need such a token?

Reply:
> Do you plan to add the token to these too?

at this time it is a no. in the current collected user stories, vault and cert manager only need this token in `NodePublishVolume` calls to authenticate.

> can we safely assume that these calls won't ever need such a token?

i didn't find a good reason to use the pod's token in the controller rpc services and `NodeStage/NodeUnstage` as they are cluster operations or early stage node operations.

Reply:
If it's used in EFS, a real persistent volume that coincidentally uses only NodePublish, I think that there is only tiny step to other volume plugins that may require NodeStage / ControllerPublish.

Reply:
I think NodeStage/ControllerPublish may add additional challenges because multiple pods can share the same volume at those operations.

Reply:
I think an explicit warning that tokens are do not cover ControllerPublish/NodeStage in the KEP (and in subsequent docs in github.com/kubernetes-csi/docs) will do for now.